rust/src/initramfs.rs

```diff
@@ -0,0 +1,179 @@
+//! Generate an "overlay" initramfs image
+
+use anyhow::{Context, Result};
+use gio::prelude::*;
+use openat::SimpleType;
+use std::collections::BTreeSet;
+use std::collections::{HashMap, HashSet};
+use std::fs;
+use std::io;
+use std::io::prelude::*;
+use std::os::unix::io::AsRawFd;
+use std::rc;
+use subprocess::Exec;
+
+fn list_files_recurse<P: glib::IsA<gio::Cancellable>>(
+    d: &openat::Dir,
+    path: &str,
+    filelist: &mut BTreeSet<String>,
+    cancellable: Option<&P>,
+) -> Result<()> {
+    // Maybe add a glib feature for openat-ext to support cancellability?  Or
+    // perhaps we should just use futures and ensure GCancellable bridges to that.
+    if let Some(c) = cancellable {
+        let _ = c.set_error_if_cancelled()?;
+    }
+    let meta = d.metadata(path).context("stat")?;
+    match meta.simple_type() {
+        SimpleType::Symlink | SimpleType::File => {}
+        SimpleType::Dir => {
+            for e in d.list_dir(path).context("readdir")? {
+                let e = e.context("readdir")?;
+                let name = e.file_name();
+                let name = match name.to_str() {
+                    Some(n) => n,
+                    None => anyhow::bail!("Invalid UTF-8 name {}", name.to_string_lossy()),
+                };
+                let subpath = format!("{}/{}", path, name);
+                list_files_recurse(&d, &subpath, filelist, cancellable)?;
+            }
+        }
+        _ => anyhow::bail!("Invalid non-regfile/symlink/directory: {}", path),
+    }
+    filelist.insert(path.to_string());
+    Ok(())
+}
+
+fn gather_filelist<P: glib::IsA<gio::Cancellable>>(
+    d: &openat::Dir,
+    input: &HashSet<String>,
+    cancellable: Option<&P>,
+) -> Result<BTreeSet<String>> {
+    let mut filelist: BTreeSet<String> = BTreeSet::new();
+    for file in input {
+        let file = match file.strip_prefix("/etc/") {
+            Some(f) => f,
+            None => anyhow::bail!("Invalid non-/etc path: {}", file),
+        };
+        list_files_recurse(&d, file, &mut filelist, cancellable)
+            .with_context(|| format!("Adding {}", file))?;
+    }
+    Ok(filelist)
+}
+
+fn generate_initramfs_overlay<P: glib::IsA<gio::Cancellable>>(
+    root: &openat::Dir,
+    files: &HashSet<String>,
+    cancellable: Option<&P>,
+) -> Result<fs::File> {
+    let filelist = {
+        let etcd = root.sub_dir("etc")?;
+        gather_filelist(&etcd, files, cancellable)?
+    };
+    let out_tmpf = std::rc::Rc::new(tempfile::tempfile()?);
+    {
+        let cmd = Exec::cmd("cpio")
+            .args(&[
+                "--create",
+                "--format",
+                "newc",
+                "--quiet",
+                "--reproducible",
+                "--null",
+            ])
+            .cwd(format!("/proc/self/fd/{}", root.as_raw_fd()))
+            | Exec::cmd("gzip").arg("-1");
```

Comment:
I was so confused by this and thought it was missing a `!` somewhere to call out to a macro. Then I saw the `impl BitOr` in the docs. Hah, so clever.